include/dmlc/logging.h

```diff
@@ -428,30 +426,42 @@ class LogMessageFatal : public LogMessage {
 #else
 class LogMessageFatal {
  public:
-  LogMessageFatal(const char* file, int line) {
-    log_stream_ << "[" << pretty_date_.HumanDate() << "] " << file << ":"
-                << line << ": ";
+  LogMessageFatal(const char *file, int line) {
+    Entry::ThreadLocal()->Init(file, line);
   }
-  std::ostringstream &stream() { return log_stream_; }
-  ~LogMessageFatal() DMLC_THROW_EXCEPTION {
+  std::ostringstream &stream() { return Entry::ThreadLocal()->log_stream; }
+  DMLC_NO_INLINE ~LogMessageFatal() DMLC_THROW_EXCEPTION {
 #if DMLC_LOG_STACK_TRACE
-    log_stream_ << "\n" << StackTrace(1, LogStackTraceLevel()) << "\n";
-#endif
-
-    // throwing out of destructor is evil
-    // hopefully we can do it here
-    // also log the message before throw
-#if DMLC_LOG_BEFORE_THROW
-    LOG(ERROR) << log_stream_.str();
+    Entry::ThreadLocal()->log_stream << "\n"
+                                     << StackTrace(1, LogStackTraceLevel())
+                                     << "\n";
 #endif
-    throw Error(log_stream_.str());
+    throw Entry::ThreadLocal()->Finalize();
```

Comment:
We moved all the logic except `throw` itself to another class, to make sure the unsafe region is minimized